src/Core/Cart/Calculator.php

```diff
@@ -200,8 +200,14 @@ public function getRowTotalWithoutDiscount()
     public function getDiscountTotal()
     {
         $amount = new AmountImmutable();
+        $isFreeShippingAppliedToAmount = false;
         foreach ($this->cartRules as $cartRule) {
+            if ($isFreeShippingAppliedToAmount && (bool) $cartRule->getRuleData()['free_shipping']) {
+                $initialShippingFees = $this->getFees()->getInitialShippingFees();
+                $amount = $amount->sub($initialShippingFees);
+            }
             $amount = $amount->add($cartRule->getDiscountApplied());
+            $isFreeShippingAppliedToAmount = (bool) $cartRule->getRuleData()['free_shipping'];
```

Comment:
I'm not sure to fully understand but if i'm not wrong, you want to keep `isFreeShippingAppliedToAmount` to true as soon as the first free shipping cart rule is applied until the end?

```suggestion
            if ((bool) $cartRule->getRuleData()['free_shipping']) {
                if ($isFreeShippingAppliedToAmount) {
                    $initialShippingFees = $this->getFees()->getInitialShippingFees();
                    $amount = $amount->sub($initialShippingFees);
                }
                $isFreeShippingAppliedToAmount = true;
            }
            $amount = $amount->add($cartRule->getDiscountApplied());
```

Reply:
That makes sense!

Reply:
Fixed in https://github.com/PrestaShop/PrestaShop/pull/18321/commits/6dd5baa266fbae61f58215040303d7d664442769
I modified the behat test in order to reproduce @atomiix usecase and indeed shipping fees were deducted multiple times. Suggested code fix fixes the usecase.